### ISSUE TITLE
Questions on drm

### ISSUE THREAD

**Issue by martin19** [emersion/drm_monitor#7]
Hi, I'm writing a screen capture application using drm to access primary framebuffer contents. This works pretty well already - drm_info helped me a lot understand the drm resource tree on different platforms - thanks for your efforts! Right now, I'm trying to listen to resolution changes in the application. My idea was polling the device tree repeatedly and querying the primary framebuffer to see if it changed. I've tested this: however on my intel integrated gpu running X11 I'm seeing the framebuffer id changes every few seconds or so - not only when the resolution changed. Do you have an idea why this is happening? Do you know of any other means (events, notifications) to query for resolution changes?

**Comment by emersion** [comment]
First, I think dri-devel would be a better place to ask this kind of questions.

Second, I'd really advise _against_ implementing screen capture via KMS. There are many issues with that -- requires root privileges, messes up synchronization, can't hold buffers for longer if encoding takes time, cursor + overlay planes + all other KMS properties need to be handled as well, etc. Instead, use the compositors' screen capture interfaces, be it the Wayland screencopy protocol or xdg-desktop-portal.

If you _really_ have no other choice, ffmpeg's kmsgrab code should be a good starting point -- but it fails at addressing all of the issues mentioned above, so it will be unreliable.

**Comment by martin19** [comment]
Thanks for your advice and the pointers. I've actually asked in dri-devel already but seems more like a busy place to discuss kernel topics I've not gotten any response there. I've started out with KMS video grabbing on Raspberry PI because I wanted the least overhead possible to capture framebuffers zero-copy piping into the hardware h264 encoder. I've got an experimental working implementation here if you are interested (https://github.com/martin19/snpserver). You are definitely right about the privileges (although I think there is something like a "capture" group). I think I might have solved the synchronization problems you mention by buffering the frames into an egl texture. After all its just the primary framebuffer in linear format I care about. I'd like to capture the cursor using KMS as well, the only problem is I cannot retrieve the hotspot coordinates. On the upside KMS is so low level that I don't need to care wether it is wayland or x11 or even consoles I need to capture.
I'll definitely have a look into the APIs you mention - thanks for that, too.
